### DISCUSSION TITLE
Exception 29

### DISCUSSION THREAD

**Discussion by MaceMoneta** [arendst/Tasmota#17101] [General]
I occasionally see this on several, otherwise idle modules:

00:00:06.082 RSL: INFO3 = {"Info3":{"RestartReason":{"Exception":29,"Reason":"Exception","EPC":["402500fb","00000000","00000000"],"EXCVADDR":"00000019","DEPC":"00000000","CallChain":["4024c44b","40250237","4026af7e","402521d7","40251dd0","40251950","40269988","40251e0b","40253d80","40253d44","40251dc0","40251d94","40251d5f","4010083d","402522cd","4023b8c0","402627bd","40262802","4023a514","40000f49","40000f49","40000f49","40000e19","40105855","4010585b","4010000d","4025b24c","4025b1fd","4010081c","40105ff5","4000050c"]},"BootCount":38}}

**Comment by MaceMoneta** [comment]
It's a custom build of 12.0.2.2, mostly to enable ping support. I have 100+ Tasmota running devices, and a few of them get an exception periodically. I didn't initially notice, because they restart without issue.

**Comment by s-hadinger** [comment]
Do you have the .map file associated with your custom build?

**Reply by MaceMoneta** [reply]
Unfortunately no. I'm trying to upgrade to current development to see if I get a recurrence. If so, I'll save the map.

**Comment by MaceMoneta** [comment]
No exception yet, but I did get a watchdog:

```
    "StatusSTK": {
        "Exception": 4,
        "Reason": "Software Watchdog",
        "EPC": [
            "40003b4d",
            "00000000",
            "00000000"
        ],
        "EXCVADDR": "00000000",
        "DEPC": "00000000"
    }
```

Map attached.
[tasmota.map.gz](https://github.com/arendst/Tasmota/files/10025052/tasmota.map.gz)

**Reply by MaceMoneta** [reply]
Additional details:

Module is a Sonoff S31.
Build repository is current as of commit 99378308f7a0ffc5b3db9dc8c43ac5e053e2ee2f.
I have 35 Sonoff S31 modules, only one has reported an issue. None of the other ~65 various modules have reported a problem, yet.

**Reply by sfromis** [reply]
Are you running the tool TasmoAdmin? In that case stop it. That tool is known to knock down borderline devices due to much higher load.

**Reply by MaceMoneta** [reply]
No TasmoAdmin, but I get status from all the Tasmota devices about once every 30 seconds (MQTT is not used).

```
    "StatusSTS": {
        "Time": "2022-11-16T14:49:44",
        "Uptime": "0T00:39:46",
        "UptimeSec": 2386,
        "Heap": 23,
        "SleepMode": "Dynamic",
        "Sleep": 50,
        "LoadAvg": 19,
        "MqttCount": 0,
        "POWER": "OFF",
        "Wifi": { 
            "AP": 1,
            "SSId": "moneta24Ghz",
            "BSSId": "B4:FB:E4:F7:D2:F1",
            "Channel": 1,
            "Mode": "11n",
            "RSSI": 76,
            "Signal": -62,
            "LinkCount": 1,
            "Downtime": "0T00:00:06"
        }   
    },      
```

**Reply by sfromis** [reply]
In that case, you have some similar tool running to poll the device for info. Try stopping that.

Polling is not a good way, as Tasmota publishes status updates to MQTT, which is the much recommended way of setting it up.

**Reply by MaceMoneta** [reply]
That can't be stopped, as my home automation depends on it. If one status command every 30 seconds can cause a watchdog, then any web command could cause a watchdog.

**Comment by s-hadinger** [comment]
Looking at the map file, the crash is in internal esp wifi code, visibly in `scan_perform` or `ieee80211_regdomain_is_active_scan`. This is unfortunately out of range of Tasmota.

I believe that the crash is not even in the Tasmota RTOS task.

I would recommend a full `Reset 2` and reconfiguration of the device. But that's just a wild guess.

Edit: wait, the map file is not for the first crash?

**Reply by MaceMoneta** [reply]
No, the map is for the watchdog. Still waiting to see if the exception 29 reoccurs.

**Comment by s-hadinger** [comment]
There is no software watchdog enabled in Tasmota, so it must come from core esp tasks. The latest stacktrace doesn't give us a hint of what is happening

**Comment by MaceMoneta** [comment]
Had a recurrence of the exception 29. Same map as above:

```
    "StatusSTK": {
        "Exception": 29,
        "Reason": "Exception",
        "EPC": [
            "40240851",
            "00000000",
            "00000000"
        ],
        "EXCVADDR": "00000000",
        "DEPC": "00000000",
        "CallChain": [
            "40101982",
            "40103033",
            "4027d8aa",
            "40237462",
            "402374c4",
            "4010194b",
            "40237462",
            "40237811",
            "4000050c",
            "40237878",
            "40239b58",
            "40203013",
            "40206fa5",
            "40221b5b",
            "40221c1d",
            "4010194b",
            "4023de79",
            "40221c87",
            "40218f99",
            "40225d73",
            "40226731",
            "4021886e",
            "4021a767",
            "4022fcbd",
            "4022fd28",
            "4010081c",
            "4023aaec",
            "40101c41"
        ]
    }
```

**Reply by s-hadinger** [reply]
Aha. It now appears that the crash is in `DhtShow(bool)` which calls `TempHumDewShow()`. Then it's unclear where the crash is coming from, but looks like maybe it's inside `ext_vsnprintf_malloc_P()` and `copyStr()`.

The main hypothesis is that `malloc()` fails and the code crashes. I will add some safeguards.

The failure of malloc() is either because the memory is exhausted, or there was a wrong calculation of the size and the calculated size of the string is insanely high.

Stay tuned

**Reply by s-hadinger** [reply]
Please try the latest version from the `development` branch. It does not solve the root cause, but it should not crash anymore.

**Reply by MaceMoneta** [reply]
I'll let it run on all the modules for a day and see what happens.

**Comment by MaceMoneta** [comment]
So far, I've gotten several issues. An exception 28:

```
    "StatusSTK": {
        "Exception": 28,
        "Reason": "Exception",
        "EPC": [
            "40261960",
            "00000000",
            "00000000"
        ],
        "EXCVADDR": "00000050",
        "DEPC": "00000000",
        "CallChain": [
            "40100230",
            "401011a9",
            "40270c33",
            "40240d46",
            "4027e099",
            "401010d0",
            "4027e099",
            "4027e0b9",
            "40240f30",
            "40240d46",
            "401011a9",
            "40240f30",
            "4000050c",
            "4026449b",
            "4027e099",
            "402404e5",
            "4027e098",
            "4027e098",
            "402404e5",
            "40240528",
            "4024fdfb",
            "4023aea0",
            "40105d9f",
            "40240c7c",
            "40240528",
            "401016c3",
            "4027a41e",
            "4010569d",
            "40237840",
            "40105282",
            "4024a400"
        ]
    }
```

Build is current as of commit 50565f4e65c6b68b0e0c50d71a2ea28fb408aeb2

[tasmota.map.gz](https://github.com/arendst/Tasmota/files/10037405/tasmota.map.gz)

Also several watchdogs. And I noticed that all modules that have power monitoring show no values (the values are correct in status10. Modules that have sensors report correct values.

![Screenshot 2022-11-17 22 17 09](https://user-images.githubusercontent.com/43620525/202609301-71c816b5-5a91-4cd6-bb1f-21d89bbf7093.png)

**Reply by sfromis** [reply]
The lack of energy monitoring values was due to a fix which has now been reverted.

The cause of mixed exceptions should usually be found on the hardware side, typically with availability of sufficient stable power, with a common problem being polling overtaxing the power supply of borderline devices.

**Reply by crontabb** [reply]
> ![Screenshot 2022-11-17 22 17 09](https://user-images.githubusercontent.com/43620525/202609301-71c816b5-5a91-4cd6-bb1f-21d89bbf7093.png)

I also don't see any data 12.2.0
On firmware 12.1.1 everything is fine

**Reply by s-hadinger** [reply]
It's fixed, used the latest in `development` branch

**Reply by sfromis** [reply]
Your screenshot there is 12.2.0.5, which is lastest often updated build, and thus not the same as 12.2.0

**Reply by sfromis** [reply]
Anyway, the development branch updates of precompiled binaries seems to be stalled, lacking the very latest updates, including the one to fix energy monitoring UI/JSON formatting.

**Reply by sfromis** [reply]
Precompiled binaries updated a few minutes ago, if you upgrade now, the values should be back

**Comment by Jason2866** [comment]
How old are the devices which generates Exceptions? Many of my Sonoff devices died after a few years.
For some i did search the reason for failing. It always was the power supply... Most of the time the caps.
Another typical failure of dying devices, disconnect from wifi and never reconnects. For a while disconnecting  from AC mains and connect does help. After a while later it is finally dead.

**Comment by MaceMoneta** [comment]
I see a new patch for the issue was merged. I'll deploy it and let you know tomorrow how it goes.

**Reply by s-hadinger** [reply]
Yes please. But considering your symptoms it looked like a faulty hardware

**Reply by MaceMoneta** [reply]
There were zero exceptions overnight (about 800 hours of device operation), so this looks good to me. Thanks!

**Comment by dfoxg** [comment]
I got the following restart error message with code 29

```
{"RestartReason":{"Exception":29,"Reason":"Exception","EPC":["4000df64","00000000","00000000"],"EXCVADDR":"00000000","DEPC":"00000000","CallChain":["40101584","4026ba6f","4026ba04","4026b9ab","4026ab04","4026ab2d","40268598","40264905","4026928f","40000f58","40268d02","40272467","40271d27","4024fc98","40000f49","40000f49","40000e19","40105b65","40105b6b","4010000d","40270c08","40270bb9","40100b38","401027ac","40105a5f","40103343","40103524","401039e6","40100b38","4000050c","40100b38"]},"BootCount":53}}
```

I'm running version 13.1.0, the device is restarting every 10-20 minutes.

Does somebody have an idea?

**Reply by RaphaelSquair** [reply]
have the same problem with version 13.1.0, anyone solved.

**Reply by gnoffer** [reply]
see also https://github.com/arendst/Tasmota/discussions/19606
